### PR TITLE
Added support for passing `Range` as a parameter to `.Exclude`.

### LINKED ISSUE #10: Add Exclude() overload that takes a Range

### PR DESCRIPTION
**Pull request by tacosontitan**
## TL;DR

This PR resolves #10 by adding support for passing `Range` as a parameter to `.Exclude`:

```cs
var samples = Enumerable.Range(1, 100).Exclude(6..94);
foreach (var sample in samples)
    Console.WriteLine(sample);
```

The expected output for the above snippet is `1 - 5` followed by `95 - 100` with each value on its own line.

## Format Adjustment

The existing ternary expressions for the `Count` properties did not match what was present in other methods in the file. As the other formats are generally cleaner I updated the `Count` properties to break nesting out:

```cs
var sample = foo.Bar
    ? foo.Bar
    : foo.Bar < 5
        ? foo.Fizz
        : foo.Buzz;
```

## Improved Documentation

Methods with access levels of `protected` or higher should have some form of documentation, even if their containing models are `private`, simply for consumer understanding. As such, I added documentation to the `Count` properties, and specified that documentation should be inherited from `IEnumerable<T>` where applicable.

##Constructor Visibility

I marked the constructors for the `private` types as `internal` since by definition they'll never be consumed outside of the assembly.

### CODE REVIEW

**Review by viceroypenguin — COMMENTED** [review]
One thing I note in all three implementations is that they assume that `!range.Start.IsFromStart` and `!range.End.IsFromStart`.

Thankfully, Collection and List patterns are easy, because [`Index.GetOffset(int length)`](https://learn.microsoft.com/en-us/dotnet/api/system.index.getoffset?view=net-8.0) exists specifically for this purpose. You can filter both `range.Start` and `range.End` through this method in the constructor before using them for the rest of the iterator.

For the `Core` method, it will require some additional evaluation. You can review the `Insert` for a glimpse at that pattern that will be required to handle it properly.

File: Source/SuperLinq/Exclude.cs [hunk 1]
```diff
@@ -46,50 +46,158 @@ public static IEnumerable<T> Exclude<T>(this IEnumerable<T> sequence, int startI
...
 
-	private sealed class ExcludeListIterator<T>(
-		IList<T> source,
-		int startIndex,
-		int count
-	) : ListIterator<T>
+	private sealed class ExcludeListIterator<T>
+		: ListIterator<T>
 	{
-		public override int Count =>
```
[the remark this answers is not in the document]
Comment (viceroypenguin) [reply Source/SuperLinq/Exclude.cs:70]: When there's a chain of ternaries, I prefer this format everywhere, yes.

[the remark this answers is not in the document]
Comment (tacosontitan) [reply Source/SuperLinq/Exclude.cs:70]: Sounds good, I just noticed it was different from other ternaries in the file so I opted for consistency since I had to update it anyway. I'll change that format back 😄

File: Source/SuperLinq/Exclude.cs [hunk 2]
```diff
@@ -102,6 +210,21 @@ private static IEnumerable<T> ExcludeCore<T>(IEnumerable<T> sequence, int startI
...
 				yield return item;
+
+			index++;
+		}
+	}
+
+	private static IEnumerable<T> ExcludeCore<T>(IEnumerable<T> sequence, Range range)
+	{
+		var index = 0;
+		var endIndex = range.Start.Value + range.End.Value;
```
[the remark this answers is not in the document]
Comment (tacosontitan) [reply Source/SuperLinq/Exclude.cs]: Not sure what I was thinking here, thanks for catching that.

**Comment by tacosontitan** [comment]
> One thing I note in all three implementations is that they assume that `!range.Start.IsFromStart` and `!range.End.IsFromStart`.

Still looking into this, just haven't had much time this week yet.

**Comment by tacosontitan** [comment]
I still need to add perf improvements and test coverage.

File: Source/SuperLinq/Exclude.cs [hunk 3]
```diff
@@ -46,50 +46,156 @@ public static IEnumerable<T> Exclude<T>(this IEnumerable<T> sequence, int startI
...
+
+		if (range.Start.Value - range.End.Value == 0)
+			return sequence;
+
+		return sequence switch
+		{
+			IList<T> list => new ExcludeListIterator<T>(list, range),
+			ICollection<T> collection => new ExcludeCollectionIterator<T>(collection, range),
+			_ => ExcludeCore(sequence, range)
+		};
```
Comment (viceroypenguin) [inline Source/SuperLinq/Exclude.cs:87]:
```suggestion
		ArgumentNullException.ThrowIfNull(sequence);

		if ((range.Start.IsFromEnd, range.End.IsFromEnd) == (false, false))
		{
			return Exclude(sequence, range.Start.Value, range.End.Value - range.Start.Value);
		}

		if (sequence.TryGetNonEnumeratedCount(out var count))
		{
			var (start, length) = range.GetOffsetAndLength(count);
			return Exclude(sequence, start, length);
		}

		return ExcludeCore(sequence, range);
```

Comment (viceroypenguin) [reply Source/SuperLinq/Exclude.cs:87]: This will defer to existing processing in as many cases as possible; it also simplifies the `ExcludeCollectionIterator` and `ExcludeListIterator` back to their original forms, because they will only get called from the `int, int` version of `Exclude`.

**Comment by tacosontitan** [comment]
I will have some time this weekend to work on this assuming I'm not still sick. I appreciate your patience with me and your assistance on it!
